[flaky test] TestDBResiliencyYugabyteScenarios: cluster teardown fails on an already-exited container

`TestDBResiliencyYugabyteScenarios/leader-master-and-tablet` failed in CI on #759 with **no failing
assertion in the test body** — the only failure is in the cluster teardown, after the scenario itself
had already passed.

- failed: https://github.com/hyperledger/fabric-x-committer/actions/runs/31490646098/job/93776119816?pr=759
- unrelated to #759's changes: it touches no `integration/runner` cluster code or `utils/testdb`. (The
  other two failures in that run *are* #759's and are being handled there.)

```
    yugabyte.go:122: stopping and removing node: sc_test_yuga_tablet_c1d55b32-...
    yugabyte.go:122:
        	Error Trace:	utils/testdb/container.go:449
        	            	utils/testdb/container.go:438
        	            	integration/runner/cluster_controller.go:106
        	            	integration/runner/yugabyte.go:122
        	            	testing.go:1317        # t.Cleanup
        	Error:      	Received unexpected error:
        	            	Container not running: b5c2725e0972212bd527220ff71844e328a0bfecf6e9e5b1fe7ee13a86f28af6
        	Test:       	TestDBResiliencyYugabyteScenarios/leader-master-and-tablet
    --- FAIL: TestDBResiliencyYugabyteScenarios/leader-master-and-tablet (410.79s)
DONE 10 tests, 2 failures in 414.497s
```

## Teardown treats an already-exited container as a test failure

`StopAndRemoveContainer` stops before removing, and the stop is fatal
([`utils/testdb/container.go:447`](https://github.com/hyperledger/fabric-x-committer/blob/main/utils/testdb/container.go#L447)):

```go
func (dc *DatabaseContainer) StopAndRemoveContainer(t *testing.T) {
	...
	dc.StopContainer(t)                                  // container.go:438
	require.NoError(t, dc.client.RemoveContainer(docker.RemoveContainerOptions{
		ID: dc.ContainerID(), Force: true,
	}))
}

// StopContainer stops db container.
func (dc *DatabaseContainer) StopContainer(t *testing.T) {
	require.NoError(t, dc.client.StopContainer(dc.ContainerID(), 10))   // container.go:449
}
```

go-dockerclient returns a typed `*docker.ContainerNotRunning` when the container has already exited
(`go-dockerclient@v1.13.2/container.go:596`), whose `Error()` is literally
`"Container not running: " + ID` — the message above. So any container that exits on its own before
teardown reaches it turns a passing test into a failing one.

## Why a node is already gone in this scenario

`stopAndRemoveCluster` walks `cc.nodes` in creation order
([`cluster_controller.go:102`](https://github.com/hyperledger/fabric-x-committer/blob/main/integration/runner/cluster_controller.go#L102)),
and `StartYugaCluster` creates all masters before all tablets, so **every master is stopped before the
first tablet**. A tserver left with no reachable master can exit by itself in that window.

`leader-master-and-tablet` makes that window much more likely: it has already removed one tablet and
the leader master during the test, so teardown starts from a 2-master/2-tablet cluster. The subtest also
ran **410.79s against a typical 75-90s**, consistent with a cluster that was already struggling.

Note this is *not* the deliberately-removed nodes being revisited —
`StopAndRemoveSingleNodeByIndex` drops them from `cc.nodes` (`cluster_controller.go:41`).

## The two teardown paths disagree

`utils/testdb` already has a tolerant version of exactly this call, for the shared unit-test container
([`database_setup.go:204`](https://github.com/hyperledger/fabric-x-committer/blob/main/utils/testdb/database_setup.go#L204)):

```go
if err := dc.client.StopContainer(dc.ContainerID(), 10); err != nil {
	log.Printf("Warning: failed to stop container %s: %v", dc.Name, err)
}
```

and `start` already handles the symmetric case with a typed check (`container.go:136-142`):

```go
var containerAlreadyRunning *docker.ContainerAlreadyRunning
if err != nil && !errors.As(err, &containerAlreadyRunning) {
	return err
}
```

So the fatal path is the odd one out.

## Suggested fix

`StopContainer` is only ever reached from `StopAndRemoveContainer` (its sole caller), so making it
tolerant cannot mask a stop failure anywhere else. Either:

1. Mirror the existing typed check — ignore `*docker.ContainerNotRunning`, and `*docker.NoSuchContainer`
   for a container removed by something else:

   ```go
   func (dc *DatabaseContainer) StopContainer(t *testing.T) {
   	t.Helper()
   	err := dc.client.StopContainer(dc.ContainerID(), 10)
   	var notRunning *docker.ContainerNotRunning
   	var noSuchContainer *docker.NoSuchContainer
   	if errors.As(err, &notRunning) || errors.As(err, &noSuchContainer) {
   		return // already gone; teardown has nothing to do
   	}
   	require.NoError(t, err)
   }
   ```

2. Or drop the stop from teardown altogether: `RemoveContainer` is already called with `Force: true`,
   which stops a running container, so the separate graceful stop buys teardown nothing.

Either way `StopAndRemoveContainer` becomes idempotent for a container in any state, matching the
`containerID == ""` guard it already has for one that was never created.

## Impact

A cleanup-only failure on a scenario whose own assertions all passed — so the reported failure points at
the wrong thing, and it blocks the PR's required check on a manual re-run. Distinct from #755 (fixed by
#758): there is no panic here, the binary survives, and the other 8 tests report normally.

--- a/utils/testdb/container.go
+++ b/utils/testdb/container.go
@@ -427,8 +427,11 @@ func (dc *DatabaseContainer) GetContainerLogs(t test.TestingT) string {
 }
 
 // StopAndRemoveContainer stops and removes the db container from the docker engine.
-// It is a no-op for a container that was never created, so a cluster teardown registered before
-// startup can run after a partial start.
+//
+// It is idempotent: a container that was never created, has already exited, or is already gone is not
+// an error, since the postcondition — the container is not running — already holds. A cluster teardown
+// therefore survives both a partial start and a node that exited on its own, as a yugabyte tserver does
+// once the masters it depends on are stopped.
 func (dc *DatabaseContainer) StopAndRemoveContainer(t *testing.T) {
 	t.Helper()
 	if dc.containerID == "" {
@@ -436,17 +439,31 @@ func (dc *DatabaseContainer) StopAndRemoveContainer(t *testing.T) {
 	}
 
 	dc.StopContainer(t)
-	require.NoError(t, dc.client.RemoveContainer(docker.RemoveContainerOptions{
+	err := dc.client.RemoveContainer(docker.RemoveContainerOptions{
 		ID:    dc.ContainerID(),
 		Force: true,
-	}))
+	})
+	var noSuchContainer *docker.NoSuchContainer
+	if errors.As(err, &noSuchContainer) {
+		return
+	}
+	require.NoError(t, err)
 	t.Logf("Container %s stopped and removed successfully", dc.ContainerID())
 }
 
-// StopContainer stops db container.
+// StopContainer stops db container. A container that has already exited, or that no longer exists, is
+// not an error — see [DatabaseContainer.StopAndRemoveContainer].
 func (dc *DatabaseContainer) StopContainer(t *testing.T) {
 	t.Helper()
-	require.NoError(t, dc.client.StopContainer(dc.ContainerID(), 10))
+	err := dc.client.StopContainer(dc.ContainerID(), 10)
+	var (
+		notRunning      *docker.ContainerNotRunning
+		noSuchContainer *docker.NoSuchContainer
+	)
+	if errors.As(err, &notRunning) || errors.As(err, &noSuchContainer) {
+		return
+	}
+	require.NoError(t, err)
 }
 
 // ContainerID returns the container ID.

--- a/utils/testdb/container_test.go
+++ b/utils/testdb/container_test.go
@@ -7,13 +7,18 @@ SPDX-License-Identifier: Apache-2.0
 package testdb
 
 import (
+	"context"
+	"fmt"
 	"sync"
 	"testing"
 	"time"
 
 	docker "github.com/fsouza/go-dockerclient"
+	"github.com/google/uuid"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+
+	"github.com/hyperledger/fabric-x-committer/utils/test"
 )
 
 // GetContainerLogs and ExecuteCommand accept a [test.TestingT] so that a polling condition can
@@ -51,6 +56,34 @@ func TestStopAndRemoveContainerWithoutContainerIsNoOp(t *testing.T) {
 	t.Parallel()
 	dc := &DatabaseContainer{Name: "sc_test_never_started"}
 
+	dc.StopAndRemoveContainer(t)
+}
+
+// A cluster node can exit on its own before teardown reaches it — a yugabyte tserver does once the
+// masters it depends on are stopped — so teardown must not fail on a container that is already gone.
+// Both halves of "already gone" are covered: exited but still present, and removed outright.
+func TestStopAndRemoveContainerIsIdempotent(t *testing.T) {
+	t.Parallel()
+	dc := &DatabaseContainer{
+		Name:         fmt.Sprintf("%s_teardown_%s", test.DockerNamesPrefix, uuid.NewString()[:8]),
+		Image:        DefaultPostgresImage,
+		DatabaseType: PostgresDBType,
+	}
+	ctx, cancel := context.WithTimeout(t.Context(), 3*time.Minute)
+	t.Cleanup(cancel)
+	dc.StartContainer(ctx, t)
+	// Reap directly rather than through the method under test, so a failure here cannot leak the
+	// container. Force removal of an already-removed container is the expected path once the test
+	// body has run.
+	t.Cleanup(func() {
+		_ = dc.client.RemoveContainer(docker.RemoveContainerOptions{ID: dc.ContainerID(), Force: true})
+	})
+
+	// Stop it behind teardown's back, as a node that exits by itself would.
+	require.NoError(t, dc.client.StopContainer(dc.ContainerID(), 10))
+	dc.StopAndRemoveContainer(t)
+
+	// And again, now that the container no longer exists at all.
 	dc.StopAndRemoveContainer(t)
 }
 

